controller.py

```diff
@@ -225,19 +229,49 @@ def find_and_get_url(links, rel, allowed_response_codes=None):
             opds_url=opds_url
         )
 
-        library.name = auth_document.get("name")
-        library.description = auth_document.get("service_description")
-
-        links = auth_document.get("links", {})
-        library.web_url = links.get("alternate", {}).get("href", None)
-        # TODO: Fetch the logo image and convert to base64 if it's a URL.
-        library.logo = links.get("logo", {}).get("href", None)
+        library.name = auth_document.title
+        library.description = auth_document.service_description
 
+        if auth_document.website:
+            library.web_url = auth_document.website.get("href")
+        else:
+            library.web_url = None
+
+        if auth_document.logo:
+            library.logo = auth_document.logo
+        elif auth_document.logo_link:
+            logo_response = do_get(auth_document.logo_link.get("href"), stream=True)
+            try:
+                image = Image.open(logo_response.raw)
+            except Exception, e:
+                return INVALID_AUTH_DOCUMENT.detailed(_("Could not read logo image %(image_url)s", image_url=auth_document.logo_link.get("href")))
+            # Convert to PNG.
+            buffer = StringIO()
+            image.save(buffer, format="PNG")
+            b64 = base64.b64encode(buffer.getvalue())
+            type = logo_response.headers.get("Content-Type") or auth_document.logo_link.get("type")
+            if type:
+                library.logo = "data:%s;base64,%s" % (type, b64)
+        else:
+            library.logo = None
+
+        if auth_document.service_area:
+            places, unknown, ambiguous = auth_document.service_area
+            if unknown or ambiguous:
+                return INVALID_AUTH_DOCUMENT.detailed(_("The authentication document has an unknown or ambiguous service area."))
+            for place in places:
+                service_area = get_one_or_create(self._db, ServiceArea,
+                                                 library_id=library.id,
+                                                 place_id=place.id)
+            for service_area in library.service_areas:
+                if service_area.place_id not in [p.id for p in places]:
```

Comment:
Build this list while you're iterating over `places` above, or at least outside the `for` loop so you don't create a nested loop.